# **Q1.1.** The model achieves **R² = 0.80**, explaining 80% of the variance. The **MAE is ~647** rentals per day. The **MAPE is 45%**, indicating relatively high percentage error due to low-count days.

In [3]:
from google.colab import files
uploaded = files.upload()

Saving bike_sharing.csv to bike_sharing.csv


In [4]:
bikes_df = pd.read_csv('bike_sharing.csv')
bikes_df.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,count
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [5]:
y = bikes_df.pop('count').values
bikes_df.pop('casual').values
bikes_df.pop('registered').values
bikes_df.pop('instant').values
bikes_df.pop('dteday').values

#X = bikes_df[['atemp']].values

X = bikes_df.values
X.shape

(731, 11)

In [6]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error as MAPE
from sklearn.preprocessing import StandardScaler
X_scal = StandardScaler().fit_transform(X)

reg = LinearRegression().fit(X_scal, y)
print(' R squared statistic: {:.3f}'.format(reg.score(X_scal, y)))
print(' Slope: {:.2f}'.format(reg.coef_[0]))
print(' Intercept: {:.2f}'.format(reg.intercept_))

 R squared statistic: 0.800
 Slope: 565.87
 Intercept: 4504.35


In [7]:
for i,j in zip(reg.coef_, bikes_df.columns):
    print('Weight: {:>7.2f} {:s}'.format(i,j))

Weight:  565.87 season
Weight: 1020.35 yr
Weight: -134.46 mnth
Weight:  -86.69 holiday
Weight:  138.36 weekday
Weight:   55.96 workingday
Weight: -332.70 weathersit
Weight:  371.14 temp
Weight:  581.91 atemp
Weight: -145.02 hum
Weight: -198.07 windspeed


In [8]:
from sklearn.linear_model import SGDRegressor
SGD = SGDRegressor(max_iter=50, tol=1e-3).fit(X_scal, y)
print(' R squared statistic: {:.3f}'.format(SGD.score(X_scal, y)))
print(' Slope: {:.2f}'.format(SGD.coef_[0]))
print(' Intercept: {:.2f}'.format(SGD.intercept_[0]))

 R squared statistic: 0.800
 Slope: 539.97
 Intercept: 4501.92


In [9]:
for i,j in zip(SGD.coef_, bikes_df.columns):
    print('Weight: {:>7.2f} {:s}'.format(i,j))

Weight:  539.97 season
Weight: 1021.21 yr
Weight: -116.68 mnth
Weight:  -93.24 holiday
Weight:  138.40 weekday
Weight:   60.58 workingday
Weight: -335.94 weathersit
Weight:  453.17 temp
Weight:  494.94 atemp
Weight: -147.99 hum
Weight: -199.51 windspeed


In [10]:
print('MAPE: {:>3.2f}'.format(MAPE(y, SGD.predict(X_scal))))
print('MAE:  {:>3.2f}'.format(mean_absolute_error(y, SGD.predict(X_scal))))

MAPE: 0.45
MAE:  647.28


# **Q1.2.** Using only temperature as the input variable results in **R² = 0.394**, significantly lower than the full model (0.80). The **MAE increases to ~1246** rentals, and **MAPE rises to around 66%**, indicating that temperature alone is insufficient to accurately predict bike demand.

In [11]:
import pandas as pd

bikes_df = pd.read_csv("bike_sharing.csv")

y = bikes_df['count'].values
X = bikes_df[['temp']].values   # тільки temp

In [12]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scal = scaler.fit_transform(X)

In [13]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

reg_temp = LinearRegression().fit(X_scal, y)

r2 = reg_temp.score(X_scal, y)
mae = mean_absolute_error(y, reg_temp.predict(X_scal))
mape = mean_absolute_percentage_error(y, reg_temp.predict(X_scal))

print("R²:", round(r2, 3))
print("MAE:", round(mae, 2))
print("MAPE:", round(mape, 3))

R²: 0.394
MAE: 1246.36
MAPE: 0.658


In [14]:
from sklearn.linear_model import SGDRegressor

sgd_temp = SGDRegressor(max_iter=1000, tol=1e-3, random_state=42).fit(X_scal, y)

r2_sgd = sgd_temp.score(X_scal, y)
mae_sgd = mean_absolute_error(y, sgd_temp.predict(X_scal))
mape_sgd = mean_absolute_percentage_error(y, sgd_temp.predict(X_scal))

print("SGD R²:", round(r2_sgd, 3))
print("SGD MAE:", round(mae_sgd, 2))
print("SGD MAPE:", round(mape_sgd, 3))

SGD R²: 0.394
SGD MAE: 1246.7
SGD MAPE: 0.659


# **Q1.3.** Using humidity as the only predictor results in **R² = 0.01**, indicating that it explains almost none of the variation in bike demand. The model performs poorly, with **MAE ≈ 1567 and MAPE ≈ 85%**, showing that humidity alone is not a useful predictor.

In [15]:
import pandas as pd

bikes_df = pd.read_csv("bike_sharing.csv")

y = bikes_df['count'].values
X = bikes_df[['hum']].values   # тільки hum

In [16]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scal = scaler.fit_transform(X)

In [17]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

reg_hum = LinearRegression().fit(X_scal, y)

r2 = reg_hum.score(X_scal, y)
mae = mean_absolute_error(y, reg_hum.predict(X_scal))
mape = mean_absolute_percentage_error(y, reg_hum.predict(X_scal))

print("R²:", round(r2, 3))
print("MAE:", round(mae, 2))
print("MAPE:", round(mape, 3))

R²: 0.01
MAE: 1567.3
MAPE: 0.851


In [18]:
from sklearn.linear_model import SGDRegressor

sgd_hum = SGDRegressor(max_iter=1000, tol=1e-3, random_state=42).fit(X_scal, y)

r2_sgd = sgd_hum.score(X_scal, y)
mae_sgd = mean_absolute_error(y, sgd_hum.predict(X_scal))
mape_sgd = mean_absolute_percentage_error(y, sgd_hum.predict(X_scal))

print("SGD R²:", round(r2_sgd, 3))
print("SGD MAE:", round(mae_sgd, 2))
print("SGD MAPE:", round(mape_sgd, 3))

SGD R²: 0.01
SGD MAE: 1567.24
SGD MAPE: 0.852


# **Q2.1.** In the partially normalised model, the **coefficient for mnth is −86.97** **(R²=0.764)**. However, this value is not directly comparable to the weather feature coefficients because the calendar variables (including mnth) are not scaled, while the weather features are in the 0–1 range. As a result, the magnitude of the mnth coefficient reflects its original scale (1–12 months) rather than its true relative importance.

In [19]:
X_all = bikes_df

In [20]:
X_all.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,count
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [22]:
import pandas as pd
from sklearn.linear_model import SGDRegressor

bikes_df = pd.read_csv("bike_sharing.csv")

y = bikes_df["count"].values

# прибрати не-ознаки + те, що не можна використовувати
X_all = bikes_df.drop(columns=["count", "casual", "registered", "instant", "dteday"])

# SGD на всіх даних (train=test як в умові)
SGD_all_raw = SGDRegressor(max_iter=1000, tol=1e-3, random_state=42)
SGD_all_raw.fit(X_all.values, y)

print("OK. X_all shape:", X_all.shape)

OK. X_all shape: (731, 11)


In [24]:
SGD_all_raw = SGDRegressor(tol=1e-3).fit(X_all, y)

In [25]:
for i,j in zip(SGD_all_raw.coef_, X_all.columns):
    print(i,j)

558.9690983913176 season
2066.2237008563975 yr
-86.93685175303104 mnth
-314.8827513949433 holiday
79.17237979768363 weekday
149.41605871124747 workingday
-858.2982562012631 weathersit
2521.5941933642894 temp
2304.730863718547 atemp
115.80542016743922 hum
-361.3499247499069 windspeed


In [32]:
print("R²:", round(SGD_all_raw.score(X_all.values, y), 3))

R²: 0.764


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but SGDRegressor was fitted with feature names
  warnings.warn(


# **Q2.2.** In the fully normalised model, the coefficient of **mnth changes to 139.45** **(R² = 0.8)**, compared to −86.94 in the partially normalised model. This shift occurs because, after full normalisation, all features are placed on the same scale (mean 0, standard deviation 1). As a result, the mnth coefficient becomes directly comparable to the other coefficients and reflects its relative contribution rather than the original 1–12 month scale.

In [27]:
X_all_scal = StandardScaler().fit_transform(X_all)

In [28]:
SGD_all_scal = SGDRegressor(tol=1e-3).fit(X_all_scal, y)
for i,j in zip(SGD_all_scal.coef_, bikes_df.columns):
    print(i,j)

552.5246052198454 instant
1018.0873708487761 dteday
-115.13652395460603 season
-88.60841628429014 yr
139.45058893941112 mnth
54.488436682131805 holiday
-331.7974841013134 weekday
455.4138531116782 workingday
500.88295224925406 weathersit
-146.14845815927285 temp
-199.31824070811763 atemp


In [33]:
print("R²:", round(SGD_all_scal.score(X_all_scal, y), 3))

R²: 0.8


# **Q2.3.** n the model that uses only the normalised temp and mnth features, the coefficient of **mnth increases to 288.50**. This occurs because, after removing other correlated seasonal variables, mnth absorbs part of their explanatory power, leading to a larger coefficient. The model achieves **R² = 0.415**, meaning it explains about 41.5% of the variance in bike demand. This shows that while temp and mnth capture some seasonal effects, they are insufficient to match the performance of the full model.

In [30]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression

bikes_df = pd.read_csv("bike_sharing.csv")

y = bikes_df["count"].values

# selecting temp і mnth only
X_two = bikes_df[["temp", "mnth"]].values

# two features normalisation
scaler = StandardScaler()
X_two_scal = scaler.fit_transform(X_two)

In [31]:
reg_two = LinearRegression().fit(X_two_scal, y)

print("R²:", round(reg_two.score(X_two_scal, y), 3))

for coef, name in zip(reg_two.coef_, ["temp", "mnth"]):
    print(f"{name:6s} {coef:10.2f}")

R²: 0.415
temp      1151.23
mnth       288.50


# **Q3.1.** **For temperature predictions**, the model performs well across all metrics. **R² = 0.688 **indicates that a substantial proportion of variance is explained. **MAE = 0.8 and RMSE = 1.0** show that prediction errors are small in absolute terms. **MAPE = 5.8%** confirms that the relative prediction error is low. **For rainfall predictions**, performance is significantly weaker. **R² = 0.231** indicates poor explanatory power. **MAE = 1.2 and RMSE = 2.608** suggest larger prediction errors, and the higher RMSE compared to MAE indicates the presence of larger individual errors. **MAPE becomes extremely large **because rainfall contains zero values, making percentage-based error unstable due to division by zero.

Overall, I learn that:

	•	R² measures how much variance is explained by the model.

	•	MAE (Mean Absolute Error) measures the average absolute difference between predicted and actual values.
It is expressed in the same units as the target variable, making it directly interpretable (e.g., degrees, millimeters). It reflects the typical size of the prediction error.

	•	RMSE (Root Mean Squared Error) measures the square root of the average squared differences between predicted and actual values.
Because the errors are squared before averaging, larger errors are penalized more heavily. RMSE is therefore more sensitive to outliers and large prediction mistakes than MAE.

	•	MAPE (Mean Absolute Percentage Error) measures the average absolute error as a percentage of the true values.
It expresses prediction error in relative terms, making it easy to interpret as a percentage. However, it becomes unstable or undefined when true values are zero, since it involves division by the actual value.

In [34]:
import numpy as np
from sklearn.metrics import r2_score, mean_absolute_error, mean_absolute_percentage_error, mean_squared_error

temp   = np.array([12,13,15,12,11,11,17,13,12,14])
t_pred = np.array([12,12,14,13,12,11,15,12,12,13])

rain   = np.array([0,0,5,7,1,1,0,8,0,4])
r_pred = np.array([0,1,4,7,1,0,0,0,1,4])

In [35]:
print("TEMP")
print("R²:", round(r2_score(temp, t_pred), 3))
print("MAE:", round(mean_absolute_error(temp, t_pred), 3))
print("MAPE:", round(mean_absolute_percentage_error(temp, t_pred), 3))
print("RMSE:", round(np.sqrt(mean_squared_error(temp, t_pred)), 3))

TEMP
R²: 0.688
MAE: 0.8
MAPE: 0.058
RMSE: 1.0


In [36]:
print("\nRAIN")
print("R²:", round(r2_score(rain, r_pred), 3))
print("MAE:", round(mean_absolute_error(rain, r_pred), 3))
print("MAPE:", round(mean_absolute_percentage_error(rain, r_pred), 3))
print("RMSE:", round(np.sqrt(mean_squared_error(rain, r_pred)), 3))


RAIN
R²: 0.231
MAE: 1.2
MAPE: 900719925474099.4
RMSE: 2.608


# **Q4.** **(3). They sometimes tell us the relative importance of the features in the model.**

Explanation:

Regression coefficients (β) indicate the change in the target variable for a one-unit change in a feature, holding other variables constant.

However, they do not always reflect relative importance because:
	•	Coefficients depend on the scale of the variables.
	•	Features measured in different units are not directly comparable.
	•	Multicollinearity can distort coefficient magnitudes.

Only when features are properly normalized and not highly correlated can coefficients meaningfully reflect relative importance.

Therefore, the correct statement is that they sometimes tell us the relative importance — not always.

# **Q5.1.** Simple Linear Regression Model:
	•	Slope (β₁) ≈ 25.14
	•	Intercept (β₀) ≈ 131.28

Interpretation

The slope (25.14) means that for each additional hour of sunshine, ice cream sales increase by approximately 25 units on average.

The intercept (131.28) represents the predicted sales when there are 0 hours of sunshine.

In [37]:
import numpy as np
from sklearn.linear_model import LinearRegression

In [38]:
x = np.array([4.3,6.9,0.0,10.4,5.2,1.8,8.0,9.2,2.1,1.2]).reshape(-1, 1)
y = np.array([224,271,130,419,230,184,362,351,196,180])

In [39]:
x_mean = np.mean(x)
y_mean = np.mean(y)

print("Mean x:", x_mean)
print("Mean y:", y_mean)

Mean x: 4.91
Mean y: 254.7


In [40]:
numerator = np.sum((x.flatten() - x_mean) * (y - y_mean))
denominator = np.sum((x.flatten() - x_mean)**2)

beta_1 = numerator / denominator

print("Slope (beta_1):", beta_1)

Slope (beta_1): 25.13738988885027


In [41]:
beta_0 = y_mean - beta_1 * x_mean
print("Intercept (beta_0):", beta_0)

Intercept (beta_0): 131.27541564574517


In [42]:
model = LinearRegression()
model.fit(x, y)

print("Sklearn slope:", model.coef_[0])
print("Sklearn intercept:", model.intercept_)

Sklearn slope: 25.13738988885026
Sklearn intercept: 131.2754156457452


# **Q5.2.** **The residual is +29.6.** This means actual sales were approximately 30 units higher than the model prediction.

Model Evaluation


The regression model shows strong predictive performance.

	•	R² = 0.941
This indicates that 94.1% of the variance in ice cream sales is explained by hours of sunshine. The relationship between sunshine and sales is therefore very strong.

	•	MAE = 18.78
On average, the model’s predictions differ from the true sales by about 19 units. This represents a relatively small error compared to the overall sales level.

	•	RMSE = 21.55
RMSE is slightly larger than MAE, meaning there are some larger errors, but they are not extreme. The difference between MAE and RMSE is moderate, indicating stable predictions without major outliers.

	•	MAPE = 0.072 (7.2%)
The model’s predictions deviate by approximately 7.2% on average in relative terms, which indicates good predictive accuracy.

⸻

Overall Conclusion

The model performs very well. Hours of sunshine are a strong predictor of ice cream sales, explaining most of the variation in the data. The error metrics confirm that prediction errors are relatively small both in absolute and percentage terms.

In [43]:
# regression coefficients
beta_0 = 131.28
beta_1 = 25.14

# given value
x_value = 8
actual_y = 362

# predicted value
y_pred = beta_0 + beta_1 * x_value

# residual
residual = actual_y - y_pred

print("Predicted value:", round(y_pred, 2))
print("Residual:", round(residual, 2))

Predicted value: 332.4
Residual: 29.6


In [44]:
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error, mean_absolute_percentage_error

# Data
x = np.array([4.3,6.9,0.0,10.4,5.2,1.8,8.0,9.2,2.1,1.2]).reshape(-1,1)
y = np.array([224,271,130,419,230,184,362,351,196,180])

# Fit model
model = LinearRegression()
model.fit(x, y)

y_pred = model.predict(x)

# Metrics
r2 = r2_score(y, y_pred)
mae = mean_absolute_error(y, y_pred)
rmse = np.sqrt(mean_squared_error(y, y_pred))
mape = mean_absolute_percentage_error(y, y_pred)

print("R²:", round(r2,3))
print("MAE:", round(mae,2))
print("RMSE:", round(rmse,2))
print("MAPE:", round(mape,3))

R²: 0.941
MAE: 18.78
RMSE: 21.55
MAPE: 0.072


▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪
✈